File: 프로그래머스/0/120866. 안전지대/안전지대.js
function solution(board) {
    let answer = board.length * board.length;
    const dangerZone = new Set();
    
    if (board.length === 1 && board[0][0] === 1) {
        answer = 0;
    } else {
        for (let x=0; x < board.length; x++) {
            for (let y=0; y < board.length; y++) {
                if (board[x][y] === 1) {
                    let target_x
                    let target_y

                    if (x-1 < 0) {
                        target_x = [x, x+1];
                    } else if (x+1 >= board.length) {
                        target_x = [x-1, x];
                    } else {
                        target_x = [x-1, x, x+1];
                    }

                    if (y-1 < 0) {
                        target_y = [y, y+1];
                    } else if (y+1 >= board.length) {
                        target_y = [y-1, y];
                    } else {
                        target_y = [y-1, y, y+1];
                    }

                    for (let i of target_x) {
                        for (let j of target_y) {
                            dangerZone.add(JSON.stringify([i, j]));
                        }
                    }
                }
            }
        }
        answer -= dangerZone.size;
    }

    return answer
}

File: 프로그래머스/0/120866. 안전지대/안전지대.rb
require 'set'

def solution(board)
    answer = board.size * board.size
    danger_zone = Set.new()
    if board.size == 1 && board.flatten.first == 1
        answer = 0;
    else
        board.each_with_index do |array,x|
            array.each_with_index do |value, y|
                if value == 1
                    target_x =
                        if x-1 < 0
                            [x, x+1]
                        elsif x+1 >= board.size
                            [x-1, x]
                        else
                            [x-1, x, x+1]
                        end

                    target_y =
                        if y-1 < 0
                            [y, y+1]
                        elsif y+1 >= board.size
                            [y-1, y]
                        else
                            [y-1, y, y+1]
                        end
                    danger_zone = danger_zone | target_x.product(target_y)
                end
            end
        end
        answer -= danger_zone.size
    end
    return answer
end
